### ISSUE TITLE
[Feature Request] Add benchmark for `Dict` and explore swiss table algorithm with SIMD lookup

### ISSUE THREAD

**Issue by jon-chuang** [modularml/mojo#3254]
### Review Mojo's priorities

- [X] I have read the [roadmap and priorities](https://docs.modular.com/mojo/roadmap.html#overall-priorities) and I believe this request falls within the priorities.

### What is your request?

While optimizing for small dictionaries is fine, I also think it might be nice to support cases which may involve large dicts.

For good perf on these, swiss table is a decent choice, implemented for Rust's HashMap and abseil's flat_hash_map. It utilizes SIMD to generate a bitmask from 7bit key prefixes for further probing of key exact match when doing a linear scan of a bucket.

### What is your motivation for this change?

perf

### Any other details?

This will likely be a last-mile issue to solve perf-wise

Let's see some benchmark numbers first

**Comment by martinvuyk** [comment]
Hi, interesting algorithm. But wouldn't it be simpler to just keep a B-tree index (of the first/last 7 hash bits) if the data becomes really big (often used in DBs) ? Though this would introduce much more memory overhead, and inserts would be kinda expensive. 

Maybe we could build it as a separate experimental type first and try it out. 

@mzaks @bethebunny you guys have been working with Dict a lot (I think),  thoughts?

**Comment by JoeLoser** [comment]
Benchmarking dict and understanding the tradeoffs of performance, memory usage, and various internal metadata-tracking algorithms would all be good things to do.  Feel free to work on this!  We're not actively working in this area at the moment.

**Comment by mzaks** [comment]
Current implementation of the dict in Mojo standard library is based on Pythons implementation. It implements a pseudo random probing algorithm and stores key value pairs together. This doesn't allow SIMD  comparison, but I think this is still not the main bottleneck, main bottleneck is the hash function. Hence I wrote a proposal https://github.com/modularml/mojo/blob/main/proposals/improved-hash-module.md and there are a few PRs to address it, which are currently blocked by some compiler issues see https://github.com/modularml/mojo/pull/2891. 

In the meantime we can implement Swiss table like dict and you can also have a look at https://github.com/mzaks/compact-dict I did a short presentation about it here https://youtu.be/3FKSlhZNdL0?si=fEshjEgpNvVUAdBu (I am the second presenter at 13:20) I would be happy if someone wants to take the time and add a SIMD based probing to it. Currently it works with simple linear probing.

**Comment by soraros** [comment]
There was always #1762. But since the discussions happens here, we can probably close that one.

**Comment by bethebunny** [comment]
I'm actively trying to shed responsibility for Dict to get more time to focus on other things 😅 We'd love to build out some more formal governance for the community. In the meantime there's lots of good thoughts here, I think y'all have some power to pave the way a bit with driving a concrete proposal. I'd be more than happy to jettison anything that currently exists for something more well motivated!

**Comment by mzaks** [comment]
After #4922 is going to be merge I can take up this task and try out some efficient alternatives for the Dict data structure @bethebunny and @JoeLoser would you be ok with that?

**Comment by JoeLoser** [comment]
> After [#4922](https://github.com/modular/modular/pull/4922) is going to be merge I can take up this task and try out some efficient alternatives for the Dict data structure [@bethebunny](https://github.com/bethebunny) and [@JoeLoser](https://github.com/JoeLoser) would you be ok with that?

Of course, I'd love that! ❤️ Mojo has a great opportunity for a very highly performant dictionary and you're an awesome person to go explore this space. We should not let the Python design choices hold us back at all here going forward.

**Comment by mzaks** [comment]
I started with the first PR #5056 adding a benchmark which compares general purpose Dict with StringDict. Based on this benchmark we can start modifying the Dict to close the gap to StringDict and maybe even get faster than StringDict

**Comment by conorbronsdon** [comment]
[v26.2](https://docs.modular.com/mojo/changelog/#v0262-2026-03-19) shipped a Swiss Table `Dict` implementation with SIMD group probing. Thanks for this discussion!
